# **Laboratorium 2**
Michał Szmuksta, gr. 7

**Wstęp**
Celem ćwiczenia było połączenie się z bazą danych przy użyciu pythona i wykonaniu zadanych zapytań do bazy. Ćwiczenie miało na celu naukę podstawowych klauzul języka SQL. Na początku zaimportowano niezbędne biblioteki i połączono się z bazą.

In [1060]:
from sqlalchemy import create_engine
import pandas as pd

db_string = "postgresql://wbauer_adb:adb2020@pgsql-196447.vipserv.org:5432/wbauer_adb"
db = create_engine(db_string)
connection_sqlalchemy = db.connect()

**Zadanie 1**

In [1061]:
df1 = pd.read_sql("SELECT COUNT(DISTINCT category_id) FROM film_category", con=connection_sqlalchemy)
df1

count
0     16

W wypożyczalni znajduje się 16 kategorii filmów.

**Zadanie 2**

In [1062]:
df2 = pd.read_sql("SELECT name FROM category ORDER BY name ASC", con=connection_sqlalchemy)
df2

name
0        Action
1     Animation
2      Children
3      Classics
4        Comedy
5   Documentary
6         Drama
7        Family
8       Foreign
9         Games
10       Horror
11        Music
12          New
13       Sci-Fi
14       Sports
15       Travel

**Zadanie 3**

In [1063]:
df3_1 = pd.read_sql("SELECT title, release_year FROM film WHERE release_year = (SELECT MAX(release_year) FROM film)", con=connection_sqlalchemy)
df3_1

title  release_year
0      Chamber Italian          2006
1     Grosse Wonderful          2006
2      Airport Pollock          2006
3    Bright Encounters          2006
4     Academy Dinosaur          2006
..                 ...           ...
995     Young Language          2006
996         Youth Kick          2006
997       Zhivago Core          2006
998  Zoolander Fiction          2006
999          Zorro Ark          2006

[1000 rows x 2 columns]

In [1064]:
df3_2 = pd.read_sql("SELECT title, release_year FROM film WHERE release_year = (SELECT MIN(release_year) FROM film)", con=connection_sqlalchemy)
df3_2

title  release_year
0      Chamber Italian          2006
1     Grosse Wonderful          2006
2      Airport Pollock          2006
3    Bright Encounters          2006
4     Academy Dinosaur          2006
..                 ...           ...
995     Young Language          2006
996         Youth Kick          2006
997       Zhivago Core          2006
998  Zoolander Fiction          2006
999          Zorro Ark          2006

[1000 rows x 2 columns]

Po wysłaniu zapytania do bazy zarówno o najstarszy jak i najmłodszy film, funkcja zwróciła dokładnie taki sam wynik, co jest równoważne z faktem, że każdy film pochodzi z tego samego roku.

**Zadanie 4**

In [1065]:
df4 = pd.read_sql("SELECT COUNT(*) FROM rental WHERE rental_date BETWEEN '2005-07-01' AND '2005-08-01'", con=connection_sqlalchemy)
df4

count
0   6709

Pomiędzy 2005-07-01 a 2005-08-01 odbyło się 6709 wypożyczeń.

**Zadanie 5**

In [1066]:
df5 = pd.read_sql("SELECT COUNT(*) FROM rental WHERE rental_date BETWEEN '2010-01-01' AND '2011-02-01'", con=connection_sqlalchemy)
df5

count
0      0

Pomiędzy 2010-01-01 a 2011-02-01 nie odbyło się żadne wypożyczenie.

**Zadanie 6**

In [1067]:
df6 = pd.read_sql("SELECT * FROM payment WHERE amount = (SELECT MAX(amount) FROM payment)", con=connection_sqlalchemy)
df6

payment_id  customer_id  staff_id  rental_id  amount  \
0       20403          362         1      14759   11.99   
1       22650          204         2      15415   11.99   
2       23757          116         2      14763   11.99   
3       24553          195         2      16040   11.99   
4       24866          237         2      11479   11.99   
5       28799          591         2       4383   11.99   
6       28814          592         1       3973   11.99   
7       29136           13         2       8831   11.99   

                payment_date  
0 2007-03-21 21:57:24.996577  
1 2007-03-22 22:17:22.996577  
2 2007-03-21 22:02:26.996577  
3 2007-03-23 20:47:59.996577  
4 2007-03-02 20:46:39.996577  
5 2007-04-07 19:14:17.996577  
6 2007-04-06 21:26:57.996577  
7 2007-04-29 21:06:07.996577

Największa płatność za wypożyczenie wynosiła 11.99 i odbyło się osiem płatności za taką cenę.

**Zadanie 7**

In [1068]:
df7 = pd.read_sql("SELECT cu.*, co.country FROM customer AS cu "
                  "JOIN address AS ad ON ad.address_id = cu.address_id "
                  "JOIN city AS ci ON ci.city_id = ad.city_id "
                  "JOIN country AS co ON co.country_id = ci.country_id "
                  "WHERE co.country IN ('Poland' ,'Nigeria' ,'Bangladesh')", con=connection_sqlalchemy)
df7

customer_id  store_id first_name  last_name  \
0           284         1      Sonia    Gregory   
1           318         1      Brian      Wyman   
2           572         1     Sidney   Burleson   
3           409         2     Rodney    Moeller   
4           332         1    Stephen     Qualls   
5           286         1      Velma      Lucas   
6            80         1    Marilyn       Ross   
7           198         2      Elsie     Kelley   
8           103         1     Gladys   Hamilton   
9           329         2      Frank   Waggoner   
10          128         1   Marjorie     Tucker   
11           18         2      Carol     Garcia   
12          270         1       Leah     Curtis   
13          501         1      Ruben      Geary   
14          277         2       Olga    Jimenez   
15          156         1     Bertha   Ferguson   
16          280         2     Tracey    Barrett   
17          250         2         Jo     Fowler   
18          571         2    Johnnie   Chisholm   
19          562         1    Wallace      Slone   
20           21         1   Michelle      Clark   
21          380         1    Russell    Brinson   
22          558         1     Jimmie  Eggleston   
23          232         2  Constance       Reid   

                                  email  address_id  activebool create_date  \
0      sonia.gregory@sakilacustomer.org         289        True  2006-02-14   
1        brian.wyman@sakilacustomer.org         323        True  2006-02-14   
2    sidney.burleson@sakilacustomer.org         578        True  2006-02-14   
3     rodney.moeller@sakilacustomer.org         414        True  2006-02-14   
4     stephen.qualls@sakilacustomer.org         337        True  2006-02-14   
5        velma.lucas@sakilacustomer.org         291        True  2006-02-14   
6       marilyn.ross@sakilacustomer.org          84        True  2006-02-14   
7       elsie.kelley@sakilacustomer.org         202        True  2006-02-14   
8    gladys.hamilton@sakilacustomer.org         107        True  2006-02-14   
9     frank.waggoner@sakilacustomer.org         334        True  2006-02-14   
10   marjorie.tucker@sakilacustomer.org         132        True  2006-02-14   
11      carol.garcia@sakilacustomer.org          22        True  2006-02-14   
12       leah.curtis@sakilacustomer.org         275        True  2006-02-14   
13       ruben.geary@sakilacustomer.org         506        True  2006-02-14   
14      olga.jimenez@sakilacustomer.org         282        True  2006-02-14   
15   bertha.ferguson@sakilacustomer.org         160        True  2006-02-14   
16    tracey.barrett@sakilacustomer.org         285        True  2006-02-14   
17         jo.fowler@sakilacustomer.org         254        True  2006-02-14   
18  johnnie.chisholm@sakilacustomer.org         577        True  2006-02-14   
19     wallace.slone@sakilacustomer.org         568        True  2006-02-14   
20    michelle.clark@sakilacustomer.org          25        True  2006-02-14   
21   russell.brinson@sakilacustomer.org         385        True  2006-02-14   
22  jimmie.eggleston@sakilacustomer.org         564        True  2006-02-14   
23    constance.reid@sakilacustomer.org         236        True  2006-02-14   

               last_update  active     country  
0  2013-05-26 14:49:45.738       1     Nigeria  
1  2013-05-26 14:49:45.738       1      Poland  
2  2013-05-26 14:49:45.738       1      Poland  
3  2013-05-26 14:49:45.738       1     Nigeria  
4  2013-05-26 14:49:45.738       1  Bangladesh  
5  2013-05-26 14:49:45.738       1     Nigeria  
6  2013-05-26 14:49:45.738       1     Nigeria  
7  2013-05-26 14:49:45.738       1     Nigeria  
8  2013-05-26 14:49:45.738       1     Nigeria  
9  2013-05-26 14:49:45.738       1  Bangladesh  
10 2013-05-26 14:49:45.738       1      Poland  
11 2013-05-26 14:49:45.738       1     Nigeria  
12 2013-05-26 14:49:45.738       1      Poland  
13 2013-05-26 14:49:45.738       1      Poland  
14 2013-05-26 14:

**Zadanie 8**

In [1069]:
df8 = pd.read_sql("SELECT staff_id, first_name, last_name, co.country, ci.city, ad.address FROM staff AS st "
                  "JOIN address AS ad ON ad.address_id = st.address_id "
                  "JOIN city AS ci ON ci.city_id = ad.city_id "
                  "JOIN country AS co ON co.country_id = ci.country_id ", con=connection_sqlalchemy)
df8

staff_id first_name last_name    country        city               address
0         1       Mike   Hillyer     Canada  Lethbridge     23 Workhaven Lane
1         2        Jon  Stephens  Australia   Woodridge  1411 Lillydale Drive

**Zadanie 9**

In [1070]:
df9 = pd.read_sql("SELECT COUNT(*) FROM staff AS st "
                  "JOIN address AS ad ON ad.address_id = st.address_id "
                  "JOIN city AS ci ON ci.city_id = ad.city_id "
                  "JOIN country AS co ON co.country_id = ci.country_id "
                  "WHERE co.country IN ('Argentina', 'Spain')", con=connection_sqlalchemy)
df9

count
0      0

W Argentynie lub Hiszpanii nie mieszka żaden pracownik.

**Zadanie 10**

In [1071]:
df10 = pd.read_sql("SELECT DISTINCT ca.name FROM rental AS re "
                  "JOIN inventory AS inv ON inv.inventory_id = re.inventory_id "
                  "JOIN film AS fi ON fi.film_id = inv.film_id "
                  "JOIN film_category AS fc ON fc.film_id = fi.film_id "
                  "JOIN category AS ca ON ca.category_id = fc.category_id ", con=connection_sqlalchemy)
df10

name
0        Sports
1      Classics
2           New
3        Family
4        Comedy
5     Animation
6        Travel
7         Music
8         Drama
9        Horror
10       Sci-Fi
11        Games
12  Documentary
13      Foreign
14       Action
15     Children

**Zadanie 11**

In [1072]:
df11 = pd.read_sql("SELECT DISTINCT ca.name FROM rental AS re "
                  "JOIN inventory AS inv ON inv.inventory_id = re.inventory_id "
                  "JOIN film AS fi ON fi.film_id = inv.film_id "
                  "JOIN film_category AS fc ON fc.film_id = fi.film_id "
                  "JOIN category AS ca ON ca.category_id = fc.category_id "
                  "JOIN customer AS cu ON cu.customer_id = re.customer_id "
                  "JOIN address AS ad ON ad.address_id = cu.address_id "
                  "JOIN city AS ci ON ci.city_id = ad.city_id "
                  "JOIN country AS co ON co.country_id = ci.country_id "
                  "WHERE co.country = 'United States'", con=connection_sqlalchemy)
df11

name
0        Action
1     Animation
2      Children
3      Classics
4        Comedy
5   Documentary
6         Drama
7        Family
8       Foreign
9         Games
10       Horror
11        Music
12          New
13       Sci-Fi
14       Sports
15       Travel

**Zadanie 12**

In [1073]:
df12 = pd.read_sql("SELECT DISTINCT fi.title FROM film AS fi "
                  "JOIN film_actor AS fa ON fa.film_id = fi.film_id "
                  "JOIN actor AS a ON a.actor_id = fa.actor_id "
                  "WHERE (a.first_name = 'Olympia' AND a.last_name = 'Pfeiffer')"
                  "OR (a.first_name = 'Julia' AND a.last_name = 'Zellweger')"
                  "OR (a.first_name = 'Ellen' AND a.last_name = 'Presley')", con=connection_sqlalchemy)
df12

title
0            Badman Dawn
1        Bilko Anonymous
2   Breakfast Goldfinger
3      Caribbean Liberty
4       Casper Dragonfly
..                   ...
64             Turn Star
65            Wait Cider
66          Women Dorado
67             Won Dares
68         Wyoming Storm

[69 rows x 1 columns]